src/core/internalMemMgr.cpp

```diff
@@ -563,4 +582,123 @@ uint32 InternalMemMgr::GetReferencesCount()
+        // Try to find the allocation in the pool list
+        for (auto it = m_poolList.Begin(); it.Get() != nullptr; it.Next())
+        {
+            GpuMemoryPool* pPool = it.Get();
+
+            PAL_ASSERT((pPool->pGpuMemory != nullptr) && (pPool->pBuddyAllocator != nullptr));
+
+            if (pPool->pGpuMemory == pGpuMemory)
+            {
+                if (pPool->pData == nullptr)
+                {
+                    result = pPool->pGpuMemory->Map(&pPool->pData);
+                    if (result != Result::Success)
+                    {
+                        pPool->pData = nullptr;
+                        break;
+                    }
+                    m_totalSizeMappedPools += pPool->pGpuMemory->Desc().size;
+                    CheckMappedPoolLimit();
+                }
+                else if (pPool->refCount == 0)
+                {
+                    // should be in unused list, remove it from there.
+                    for (auto it2 = m_unusedMappedPoolList.Begin(); it2.Get() != nullptr; it2.Next())
+                    {
+                        if (*(it2.Get()) == pPool)
+                        {
+                            m_unusedMappedPoolList.Erase(&it2);
+                            break;
+                        }
+                    }
+                }
+                pPool->refCount++;
+                *ppData = pPool->pData;
+                result = Result::Success;
+                break;
+            }
+        }
+
+        // If we didn't find the allocation in the pool list then something went wrong with the allocation scheme
+        PAL_ASSERT(result == Result::Success);
+    }
+    else
+    {
+        result = pGpuMemory->Map(ppData);
+    }
+
+    return result;
+}
+
+Result InternalMemMgr::Unmap(
+        GpuMemory* pGpuMemory)
+{
+    PAL_ASSERT(pGpuMemory != nullptr);
+    if (pGpuMemory->WasBuddyAllocated())
+    {
+        Util::MutexAuto allocatorLock(&m_allocatorLock); // Ensure thread-safety using the lock
+        // Try to find the allocation in the pool list
+        for (auto it = m_poolList.Begin(); it.Get() != nullptr; it.Next())
+        {
+            GpuMemoryPool* pPool = it.Get();
+
+            PAL_ASSERT((pPool->pGpuMemory != nullptr) && (pPool->pBuddyAllocator != nullptr));
+            if (pPool->pGpuMemory == pGpuMemory)
+            {
+                if (pPool->pData != nullptr)
+                {
+                    pPool->refCount--;
+                    if (pPool->refCount == 0)
+                    {
+                        m_unusedMappedPoolList.PushBack(pPool);
+                        CheckMappedPoolLimit();
+                    }
+                }
+                break;
+            }
+        }
+    }
+    else
+    {
+        pGpuMemory->Unmap();
+    }
+
+    return Result::Success;
+}
+
+void InternalMemMgr::CheckMappedPoolLimit()
+{
+    if ((m_pDevice->Settings().maxMappedPoolsSize >= 0))
```

Comment:
FYI:  It looks like you accidentally put two sets of parentheses around this conditional.

Reply:
Done